{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e2832ab9",
   "metadata": {},
   "source": [
    "## Overview\n",
    "\n",
    "Welcome to the first notebook of the fba workshops! The learning objectives of this notebook are as follows:\n",
    "\n",
    "- Understanding flux-balance analysis\n",
    "- Understanding the use of GEMs/GSMMs in flux-balance analysis\n",
    "- Getting familiar with the COBRApy framework\n",
    "    - Importing a GSMM\n",
    "    - Finding reactions & metabolites\n",
    "    - Objective functions and running the analysis\n",
    "    - Altering reaction upper and lower bounds\n",
    "\n",
    "The take-home practice section will then try to give you intuition on:\n",
    "\n",
    "- hybrid objective functions\n",
    "- competing pathways\n",
    "- knock-outs\n",
    "\n",
    "**Before getting started:** Please put the names of your group members into the markdown cell below so that you can take credit for your work."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5e67002",
   "metadata": {},
   "source": [
    "**Group Members:**\n",
    "\n",
    "*Alexis Moutafis-Tymcio*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e37486",
   "metadata": {},
   "source": [
    "## What is flux-balance analysis?\n",
    "\n",
    "Flux-balance analysis is a predictive modeling method that is significantly used in systems biology. It has a wide variety of applications including (but not limited to):\n",
    "\n",
    "1. Media optimization for the culture\n",
    "2. Finding knock-out targets\n",
    "3. Understanding metabolic constraints\n",
    "4. Neuroscience\n",
    "5. Biomanufacturing\n",
    "6. Cancer Research\n",
    "\n",
    "Many iGEM teams have cited [this paper](https://www.nature.com/articles/nbt.1614) to explain how flux-balance analysis works. In a nutshell, it derives from the idea that reactions and metabolites in a metabolism can be represented as a matrix by the following **example:** \n",
    "\n",
    "Let's take an arbitrary reaction:\n",
    "\n",
    "$$2Na + 2HCl \\rightarrow 2NaCl + H_2$$\n",
    "\n",
    "This reaction can be represented as a column vector:\n",
    "\n",
    "$$v = \\begin{pmatrix} -2 & -2 & 2 & 1 \\end{pmatrix}^T$$\n",
    "\n",
    "Where each item represents the coefficient consumed/produced of that metabolite in the reaction. This principle applied to thousands of metabolites and reactions, and we get a linear optimization problem formulated as $Sv = 0$ where:\n",
    "\n",
    "- $S \\in R^{M \\times N}$ represents the stoichiometric matrix with each column a reaction, and each row a metabolite.\n",
    "- $v \\in R^{N}$ represents the fluxes (rates) of reactions, usually in units $mmol/gDW/h$ (biomass reactions are usually in $1/h$). \n",
    "- an assumption is made that the metabolic network is at steady-state, where each metabolite is consumed and produced at the same amount.\n",
    "\n",
    "Given a metabolite concentration $[x]$ with the corresponding coefficient $s_i$ in the $i^{th}$ reaction, and $t$ for time, an equivalent formulation of the problem would be $$ \\sum_{i=1}^N s_i v_i = \\frac{d[x]}{dt} = 0 $$\n",
    "\n",
    "For more advanced analyses, flux-balance analysis pairs with [multiomics data](https://en.wikipedia.org/wiki/Multiomics) which can introduce more information such as genetics, transcriptomics, epigenetics, etc.\n",
    "\n",
    "### What are genome-scale metabolic models?\n",
    "\n",
    "Genome-scale metabolic models (GSMM or GEM for short) are formatted data files that contains all the essential information of the metabolism \n",
    "of an organism. Not only the set of reactions and metabolites, but they hold much more information such as gene associations (for turning on/off phenotypes), model compartments, flux constraints, etc. Development of comprehensive and accurate GSMMs for non-model organisms is one of the greatest bottlenecks of systems biology.\n",
    "\n",
    "Most GSMMs are published and used in the file format called SBML, which stands for **systems biology markup language**. However, the JSON file format, originally created for data transfer in web software applications, is also used for a similar purpose.\n",
    "\n",
    "### The COBRA Toolbox\n",
    "\n",
    "The [iJO1366](http://bigg.ucsd.edu/models/iJO1366) model was developed to mimic the metabolism of *Escherichia coli*. In the following practice, you'll be introduced to the basics of the python library COBRApy that is used to compile these models and run very basic forms of flux-balance analysis.\n",
    "\n",
    "First start off by installing COBRApy and making necessary imports into the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "39ec458c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: cobra in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (0.30.0)\n",
      "Requirement already satisfied: appdirs~=1.4 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (1.4.4)\n",
      "Requirement already satisfied: depinfo~=2.2 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (2.2.0)\n",
      "Requirement already satisfied: diskcache~=5.0 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (5.6.3)\n",
      "Requirement already satisfied: future in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (1.0.0)\n",
      "Requirement already satisfied: httpx~=0.24 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (0.27.0)\n",
      "Requirement already satisfied: numpy>=1.13 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (2.0.2)\n",
      "Requirement already satisfied: optlang~=1.8 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (1.8.3)\n",
      "Requirement already satisfied: pandas<3.0,>=1.0 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (2.2.2)\n",
      "Requirement already satisfied: pydantic>=1.6 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (2.7.1)\n",
      "Requirement already satisfied: python-libsbml~=5.19 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (5.21.0)\n",
      "Requirement already satisfied: rich>=8.0 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (13.7.1)\n",
      "Requirement already satisfied: ruamel.yaml~=0.16 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (0.19.1)\n",
      "Requirement already satisfied: swiglpk in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from cobra) (5.0.13)\n",
      "Requirement already satisfied: anyio in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from httpx~=0.24->cobra) (3.7.1)\n",
      "Requirement already satisfied: certifi in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from httpx~=0.24->cobra) (2021.5.30)\n",
      "Requirement already satisfied: httpcore==1.* in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from httpx~=0.24->cobra) (1.0.5)\n",
      "Requirement already satisfied: idna in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from httpx~=0.24->cobra) (3.2)\n",
      "Requirement already satisfied: sniffio in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from httpx~=0.24->cobra) (1.2.0)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from httpcore==1.*->httpx~=0.24->cobra) (0.14.0)\n",
      "Requirement already satisfied: sympy>=1.12.0 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from optlang~=1.8->cobra) (1.12)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from pandas<3.0,>=1.0->cobra) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from pandas<3.0,>=1.0->cobra) (2021.3)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from pandas<3.0,>=1.0->cobra) (2023.3)\n",
      "Requirement already satisfied: annotated-types>=0.4.0 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from pydantic>=1.6->cobra) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.18.2 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from pydantic>=1.6->cobra) (2.18.2)\n",
      "Requirement already satisfied: typing-extensions>=4.6.1 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from pydantic>=1.6->cobra) (4.12.0)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from python-dateutil>=2.8.2->pandas<3.0,>=1.0->cobra) (1.16.0)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from rich>=8.0->cobra) (3.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from rich>=8.0->cobra) (2.18.0)\n",
      "Requirement already satisfied: mdurl~=0.1 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from markdown-it-py>=2.2.0->rich>=8.0->cobra) (0.1.2)\n",
      "Requirement already satisfied: mpmath>=0.19 in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from sympy>=1.12.0->optlang~=1.8->cobra) (1.3.0)\n",
      "Requirement already satisfied: exceptiongroup in c:\\users\\alexi\\appdata\\local\\programs\\python\\python39\\lib\\site-packages (from anyio->httpx~=0.24->cobra) (1.2.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install cobra\n",
    "\n",
    "from cobra import io, Model, Solution\n",
    "import argparse, os, sys"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bd8cb6c",
   "metadata": {},
   "source": [
    "Now, we read our model file into a `Cobra.Model` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d2b10c7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 metabolites: \n",
      "(1) ID: 10fthf_c - Name: 10-Formyltetrahydrofolate\n",
      "(2) ID: 12dgr120_c - Name: 1,2-Diacyl-sn-glycerol (didodecanoyl, n-C12:0)\n",
      "(3) ID: 12dgr140_c - Name: 1,2-Diacyl-sn-glycerol (ditetradecanoyl, n-C14:0)\n",
      "(4) ID: 12dgr141_c - Name: 1,2-Diacyl-sn-glycerol (ditetradec-7-enoyl, n-C14:1)\n",
      "(5) ID: 12dgr160_c - Name: 1,2-Diacyl-sn-glycerol (dihexadecanoyl, n-C16:0)\n",
      "\n",
      "First 5 reactions: \n",
      "(1) ID: EX_cm_e - Name: Chloramphenicol exchange\n",
      "(2) ID: EX_cmp_e - Name: CMP exchange\n",
      "(3) ID: EX_co2_e - Name: CO2 exchange\n",
      "(4) ID: EX_cobalt2_e - Name: Co2+ exchange\n",
      "(5) ID: DM_4crsol_c - Name: Sink needed to allow p-Cresol to leave system\n"
     ]
    }
   ],
   "source": [
    "model: Model = io.read_sbml_model('./iJO1366.xml')\n",
    "\n",
    "five_mets = [f'\\n({i+1}) ID: {met.id} - Name: {met.name}' for i, met in enumerate(model.metabolites[:5])]\n",
    "five_rxns = [f'\\n({i+1}) ID: {rxn.id} - Name: {rxn.name}' for i, rxn in enumerate(model.reactions[:5])]\n",
    "print(f\"First 5 metabolites: {''.join(five_mets)}\\n\")\n",
    "print(f\"First 5 reactions: {''.join(five_rxns)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b91217e3",
   "metadata": {},
   "source": [
    "The above code reads the `sbml` file into the `Model` object, and then reads out the names and ids of the first 5 reactions and metabolites in the array. To understand how data is stored in the `Model` object, we can further look into the `Metabolite` and `Reaction` classes, some of which we'll cover in this notebook. If you're curious, you can visit the [COBRApy documentation page](https://cobrapy.readthedocs.io/en/latest/) for more information.\n",
    "\n",
    "Here's a neater way to display the model information on jupyter notebook, which neatly renders as a table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c1b14a3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iJO1366</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>25d37619610</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1805</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2583</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1367</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_Ec_iJO1366_core_53p95M - 1.0*BIOMASS_Ec_iJO1366_core_53p95M_reverse_5c8b1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iJO1366 at 0x25d37619610>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c6c9e8",
   "metadata": {},
   "source": [
    "As previously discussed, flux-balance analysis makes a major assumption; that all metabolites are produced and consumed at the same amount. This assumption describes a steady-state. Within this assumption, we're trying to find an arrangement of flux vector $v$ that gives the highest (maximum) flux through our target reaction (objective).\n",
    "\n",
    "Let's assume we want to optimize for the **biomass** objective function. The name of this function may vary based on the GSMM and the database that was used to construct it, but the **biomass** keyword is usually in there somewhere.\n",
    "\n",
    "With the following script, we can identify the biomass reaction by looking for the **biomass** keyword in all reaction IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2ecc00e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biomass Reactions: \n",
      "- BIOMASS_Ec_iJO1366_WT_53p95M: E. coli biomass objective function (iJO1366) - WT - with 53.95 GAM estimate\n",
      "- BIOMASS_Ec_iJO1366_core_53p95M: E. coli biomass objective function (iJO1366) - core - with 53.95 GAM estimate\n"
     ]
    }
   ],
   "source": [
    "biomass_rxns = [rxn for rxn in model.reactions if 'biomass' in rxn.id.lower()]\n",
    "\n",
    "print(f'Biomass Reactions: ')\n",
    "for rxn in biomass_rxns:\n",
    "    print(f'- {rxn.id}: {rxn.name}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8fa9c7",
   "metadata": {},
   "source": [
    "Some models may have multiple biomass functions. Usually different biomass functions indicate \n",
    "\n",
    "1. the estimations via different experiments, or \n",
    "2. predictions under different medium conditions, representing different \"modes\" of the organism.\n",
    "\n",
    "As an example, in our [iGEM 2025 (Sterosaurus)](https://2025.igem.wiki/mcmasteru/) project, we used the iCre1355 model of the *C. reinhardtii* algae organism. There were 3 biomass functions representing the autotrophic, mixotrophic, and heterotrophic mode. These modes represent the growth function of the organism under different medium conditions i.e. O2 stress, CO2 stress, etc. You can find the paper for the GSMM model we used in our project [here](https://onlinelibrary.wiley.com/doi/10.1111/tpj.13059).\n",
    "\n",
    "For our optimization purposes, we'll use the core biomass function `BIOMASS_Ec_iJO1366_core_53p95M`. Let's now optimize our model and get the flux results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2a3dd27d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 5 fluxes in optimal solution:\n",
      "ATPS4rpp      55.815247\n",
      "EX_h2o_e      45.619430\n",
      "CYTBO3_4pp    35.149539\n",
      "NADH16pp      31.021723\n",
      "EX_co2_e      19.675223\n",
      "Name: fluxes, dtype: float64\n",
      "\n",
      "Optimal growth rate: 0.9823718127269743 1/hr\n"
     ]
    }
   ],
   "source": [
    "biomass_obj = 'BIOMASS_Ec_iJO1366_core_53p95M'\n",
    "\n",
    "model.objective = biomass_obj\n",
    "solution: Solution = model.optimize()\n",
    "\n",
    "solution.fluxes.sort_values(ascending=False, inplace=True)\n",
    "print(f'Top 5 fluxes in optimal solution:')\n",
    "print(solution.fluxes.head(5))\n",
    "\n",
    "print(f'\\nOptimal growth rate: {solution.objective_value} 1/hr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf0848e3",
   "metadata": {},
   "source": [
    "Now, moving on to different features.\n",
    "\n",
    "Additionally, reactions can be upper and lower bounded within the model. We can do this by reaching a reaction of our choice within the model and change its `upper_bound` and `lower_bound` parameters. Let's try to do this on the medium uptake of our model.\n",
    "\n",
    "To find a reaction of interest, we may think we need to iterate through every reaction in the model. However, there's simple ways in which we can reduce the workload. For instance, the cobra `Model` object is known to keep record of the compartments of each metabolite. We can first write a script to print out the compartments available in the model, and then filter only for the exchange reactions that have extracellular metabolites as reactants.\n",
    "\n",
    "We start by listing out the compartments of our model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "379e387f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'c': 'cytosol', 'e': 'extracellular space', 'p': 'periplasm'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.compartments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bfd66a0",
   "metadata": {},
   "source": [
    "Notice that the `'e': 'extracellular space'` compartment fits our use case. We're looking for the extracellular oxygen metabolite to find its relevant reactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5c1c33d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('co2_e', 'CO2 CO2'),\n",
       " ('h2o2_e', 'Hydrogen peroxide'),\n",
       " ('kdo2lipid4_e', 'KDO(2)-lipid IV(A)'),\n",
       " ('so2_e', 'Sulfur dioxide'),\n",
       " ('no2_e', 'Nitrite'),\n",
       " ('o2_e', 'O2 O2'),\n",
       " ('o2s_e', 'Superoxide anion')]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Query for all metabolites with 'o2' in their ID or name\n",
    "mets = model.metabolites.query('.*o2.*')\n",
    "# Print only those in the extracellular compartment\n",
    "output = []\n",
    "for met in mets:\n",
    "    if met.compartment != 'e':\n",
    "        continue\n",
    "    output.append((met.id, met.name))\n",
    "\n",
    "output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e900728",
   "metadata": {},
   "source": [
    "The sixth item seems to be the metabolite we're loooking for. Now, let's try to constrain the model based on its oxygen uptake. We know that the oxygen uptake reaction must contain our metabolite as a reactant. So let's find the oxygen uptake reaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ee506ae4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- EX_o2_e: O2 exchange: -1000.0 to 1000.0 mmol/gDW/hr\n",
      "- O2tex: Oxygen transport via diffusion (extracellular to periplasm): -1000.0 to 1000.0 mmol/gDW/hr\n"
     ]
    }
   ],
   "source": [
    "o2e_id, o2e_name = output[5]\n",
    "\n",
    "rxns = []\n",
    "for rxn in model.reactions:\n",
    "    if o2e_id in [met.id for met, coef in rxn.metabolites.items() if coef < 0]:\n",
    "        rxns.append(rxn)\n",
    "\n",
    "for r in rxns:\n",
    "    print(f'- {r.id}: {r.name}: {r.lower_bound} to {r.upper_bound} mmol/gDW/hr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d31549b",
   "metadata": {},
   "source": [
    "The above code filters the reactions that include our metabolite as a reactant, and prints out their id, name, and constraints. In the context of a metabolic model, a range of -1000 to 1000 as upper and lower bounds is practically infinite. Let's set this much lower. The reason why we take the copy of the original model in the below code is because `Model` and `Reaction` behave as reference types, and the variable `oxygen_uptake_rxn` can mutate the original object in the model. Making a copy prevents us from overwriting the original data.\n",
    "\n",
    "In the below cell block, we use the `upper_bound` and `lower_bound` properties of the `Reaction` object to constrain its activity. Upon the update, we run the analysis using the `optimize()` method of our model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9f52c917",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Optimal growth rate: 0.9823718127269705 1/hr\n",
      "\n",
      "Optimal growth rate: 0.49154514798795895 1/hr\n"
     ]
    }
   ],
   "source": [
    "# Original optimization\n",
    "solution = model.optimize()\n",
    "\n",
    "print(f'\\nOptimal growth rate: {solution.objective_value} 1/hr')\n",
    "\n",
    "# With new boundaries\n",
    "copy_model = model.copy()\n",
    "oxygen_uptake_rxn = copy_model.reactions.get_by_id('EX_o2_e')\n",
    "oxygen_uptake_rxn.lower_bound = -5.0\n",
    "oxygen_uptake_rxn.upper_bound = 5.0 \n",
    "\n",
    "solution = copy_model.optimize()\n",
    "\n",
    "print(f'\\nOptimal growth rate: {solution.objective_value} 1/hr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e368a949",
   "metadata": {},
   "source": [
    "Now that we've covered some of the basics of COBRApy, we can get started with some take-home practice questions.\n",
    "\n",
    "## Take-Home Practice\n",
    "\n",
    "#### Part 1: Hybrid Objective for Optimization\n",
    "\n",
    "In the above section, recall that we could set the objective function of a model by calling `model.objective = 'reaction_id'`. However, there's an alternative way in which we can define the model objective. If you read the documentation for the `Reaction` object, you'll see that one of its many properties is called `objective_coefficient` which is a float integer. This way, we can include multiple reactions in the objective function. It is important to note that all of the objective coefficients should add up to 1.\n",
    "\n",
    "A demonstration:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c2fbfdcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>0.5 EX_glc__D_e + 0.5 EX_o2_e = -0.429545454545454</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>0.2864</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>0.5727</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e</td>\n",
       "      <td>-0.5727</td>\n",
       "      <td>2</td>\n",
       "      <td>66.67%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-0.5727</td>\n",
       "      <td>1</td>\n",
       "      <td>33.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-0.5727</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.5727</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x25d4a132f70>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = io.read_sbml_model('./iJO1366.xml')\n",
    "\n",
    "# Reset objective coefficients\n",
    "for rxn in model.reactions:\n",
    "    rxn.objective_coefficient = 0.0\n",
    "\n",
    "# Set new objective\n",
    "model.reactions.get_by_id('EX_glc__D_e').objective_coefficient = 0.5\n",
    "model.reactions.get_by_id('EX_o2_e').objective_coefficient = 0.5\n",
    "\n",
    "\n",
    "model.optimize()\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3e73f1e",
   "metadata": {},
   "source": [
    "In the cell below, your task is to:\n",
    "\n",
    "<input type=\"checkbox\" style=\"margin-bottom: 20px;\"><label>Import the model for *Saccharomyces cerevisiae* (file named `yeast-GEM.xml`) into COBRA</label></input><br>\n",
    "<input type=\"checkbox\" style=\"margin-bottom: 20px;\"><label>Set a custom objective function where: (1) thiazole phosphate synthesis (`\"id: r_2070\"`) objective coefficient is set to 0.5 and (2) biomass function (`\"id: r_4041\"`) coefficient set to 0.5</label></input><br>\n",
    "<input type=\"checkbox\"><label>Optimize the model and print out the resulting fluxes for the two reactions. (<b>Hint:</b> Use `solution.fluxes[['rid1','rid2']]` to print results)</label></input><br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab2153e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r_2070    1.004973e-07\n",
      "r_4041    8.374771e-02\n",
      "Name: fluxes, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "model2: Model = io.read_sbml_model('./yeast-GEM.xml')\n",
    "\n",
    "# Reset objective coefficients\n",
    "for rxn in model2.reactions:\n",
    "    rxn.objective_coefficient = 0.0\n",
    "\n",
    "# Set new objective\n",
    "model2.reactions.get_by_id('r_2070').objective_coefficient = 0.5\n",
    "model2.reactions.get_by_id('r_4041').objective_coefficient = 0.5\n",
    "\n",
    "solution = model2.optimize()\n",
    "print(solution.fluxes[['r_2070', 'r_4041']])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98692dc0",
   "metadata": {},
   "source": [
    "#### Part 2: Gene Knock-out\n",
    "\n",
    "We can simulate the response of the metabolism upon knocking out different genes. We can do this with the `Gene` object in the library. We can access the gene information and knock it out by accessing the `gene_reaction_rule` property of the `Reaction` object, which returns an identifier string. We can then query that string using `model.genes.get_by_id('my_gene_rule_id')`, save it in a variable called `gene` and simply run `gene.knock_out()` to knock it out.\n",
    "\n",
    "<br>\n",
    "\n",
    "To carry out this step, you'll need to perform the following steps on the same yeast metabolism from part 1:\n",
    "\n",
    "<input type=\"checkbox\" style=\"margin-bottom: 20px;\"><label>Re-set the objective coefficients of the model (or simply import it again)</label></input><br>\n",
    "<input type=\"checkbox\" style=\"margin-bottom: 20px;\"><label>Make 2 copies of the model (for 2 different knock-out experiments)</label></input><br>\n",
    "<input type=\"checkbox\" style=\"margin-bottom: 20px;\"><label>On the first copy, knock out the pyruvate kinase reaction from its gene. The reaction has `id: r_0962`.</label></input><br>\n",
    "<input type=\"checkbox\" style=\"margin-bottom: 20px;\"><label>On the second copy, knock out the acetyl-CoA carboxylase from its gene. The reaction has `id: r_0109`.</label></input><br>\n",
    "<input type=\"checkbox\" style=\"margin-bottom: 20px;\"><label>Set the objective function to `id: r_4041` (for biomass) on the original model, and the two copies (3 models in total).</label></input><br>\n",
    "<input type=\"checkbox\"><label>Run flux-balance analysis on each of the 3 models and print out the flux of the objective reaction for clear comparison.</label></input><br>\n",
    "\n",
    "<br>\n",
    "\n",
    "**Please note that a reaction can be regulated by multiple genes.**\n",
    "\n",
    "<br>\n",
    "\n",
    "Upon completion, answer the follow-up questions:\n",
    "\n",
    "1. Are the flux values of `r_4041` the same between the original model and the first copy? What does this difference tell us about the relation of the knocked out gene to the biomass objective?\n",
    "2. Are the flux values of `r_4041` the same between the original model and  the second copy? What does this difference tell us about the relation of the knocked out gene to the biomass objective?\n",
    "3. If you mess around in COBRApy, you may notice that we can also perform knock-out on the reaction itself by setting its upper and lower bounds to 0 or simply running `reaction.knock_out()`. This seems as an easier approach to perform a knock-out. Do you think we should use this method for knock-outs instead of the gene rule? Can you explain why (or why not) this is a good (or bad) approach?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "04ab11af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YAL038W or YOR347C\n",
      "YDL141W and YNR016C\n",
      "\n",
      "\n",
      "r_4041    0.083748\n",
      "Name: fluxes, dtype: float64\n",
      "r_4041    0.07982\n",
      "Name: fluxes, dtype: float64\n",
      "r_4041    1.285378e-18\n",
      "Name: fluxes, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# Reset objective coefficients\n",
    "for rxn in model2.reactions:\n",
    "    rxn.objective_coefficient = 0.0\n",
    "\n",
    "model_pk_ko = model2.copy()\n",
    "model_acc_ko = model2.copy()\n",
    "\n",
    "#Pyruvate kinase KO\n",
    "rxn_pk = model_pk_ko.reactions.get_by_id('r_0962')\n",
    "rule_pk = rxn_pk.gene_reaction_rule\n",
    "print(rule_pk)\n",
    "\n",
    "for gene in rxn_pk.genes:\n",
    "    gene.knock_out()\n",
    "\n",
    "#Acetyl-CoA KO\n",
    "rxn_acc = model_acc_ko.reactions.get_by_id('r_0109')\n",
    "rule_acc = rxn_acc.gene_reaction_rule\n",
    "print(rule_acc)\n",
    "\n",
    "for gene in rxn_acc.genes:\n",
    "    gene.knock_out()\n",
    "\n",
    "#Objective functions\n",
    "model2.reactions.get_by_id('r_4041').objective_coefficient = 0.5\n",
    "model_pk_ko.reactions.get_by_id('r_4041').objective_coefficient = 0.5\n",
    "model_acc_ko.reactions.get_by_id('r_4041').objective_coefficient = 0.5\n",
    "\n",
    "solution = model2.optimize()\n",
    "solution_pk = model_pk_ko.optimize()\n",
    "solution_acc = model_acc_ko.optimize()\n",
    "\n",
    "print(\"\\n\")\n",
    "\n",
    "print(solution.fluxes[['r_4041']])\n",
    "print(solution_pk.fluxes[['r_4041']])\n",
    "print(solution_acc.fluxes[['r_4041']])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "759d42d2",
   "metadata": {},
   "source": [
    "**Answer to 1:** There was a relatively minimal decrease in flux when we knocked out pyruvate kinase. This would imply that it is not required for biomass production, or other pathways exist."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33364919",
   "metadata": {},
   "source": [
    "**Answer to 2:** There was a very stark decrease (basically 0) in flux when we knocked out acetyl-CoA. This implies that this compound is critical for biomass production."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afe90a02",
   "metadata": {},
   "source": [
    "**Answer to 3:** I don't think that's a good approach to take. By knocking out the entire reaction, and not a specific gene, you over-generalize (i.e. for an \"or\" reaction, knocking out one gene could still let the reaction happen). This is too broad of an assumption and could create inaccurate results."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be100268",
   "metadata": {},
   "source": [
    "In this notebook, we've explored what flux-balance analysis is, and how we can use COBRApy to load GSMMs from `sbml` files and optimize for a certain objective function.\n",
    "\n",
    "In most applications of flux-balance analysis, it is valuable to read the paper of the GSMMs model release as it contains important information relevant to our purpose.\n",
    "\n",
    "In the upcoming activities, we'll expand on the practice we did here and introduce new concepts such as flux-variability analysis, knock-out targets, troubleshooting and validating models.\n",
    "\n",
    "### Supplementary Material\n",
    "\n",
    "As previously mentioned, feel free to check out the [COBRApy documentation](https://cobrapy.readthedocs.io/en/latest/) on the use of the library.\n",
    "\n",
    "If you're curious, you can also check out [our software repository](https://gitlab.igem.org/2025/software-tools/mcmasteru) from iGEM 2025, where the flux-balance analysis code can be found under the `fba-codebase` folder. Alternatively, you can also find the code on this [GitHub repo](https://github.com/FarukEfe/McMasterU-iGEM-2025-Modelling).\n",
    "\n",
    "If you're looking for more resources Dr. Francisco Zorrilla from Cambridge University developed a set of exercises for his Part III Systems Biology course. You can check out his resources [here](https://github.com/franciscozorrilla/systems-biology-fba-practical)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
