{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "15e37486",
   "metadata": {},
   "source": [
    "### What is flux-balance analysis?\n",
    "\n",
    "Flux-balance analysis is a computational biology method that is significantly used in systems biology. It has a wide variety of applications including but not limited to:\n",
    "\n",
    "1. Culture medium optimization,\n",
    "2. Identifying gene knockouts (competitive pathways),\n",
    "3. Metabolic Engineering,\n",
    "4. Neuroscience.\n",
    "\n",
    "Many iGEM teams have cited [this paper](https://www.nature.com/articles/nbt.1614) to explain how flux-balance analysis works. In a nutshell, it derives from the idea that reactions and metabolites in a metabolism can be represented as a matrix as follows:\n",
    "\n",
    "Let's take an arbitrary reaction, for example:\n",
    "\n",
    "$$2Na + 2HCl \\rightarrow 2NaCl + H2$$\n",
    "\n",
    "This reaction can be represented as a column vector:\n",
    "\n",
    "$v = \\begin{pmatrix} -2 \\\\\\\\ -2 \\\\\\\\ 2 \\\\\\\\ 1 \\end{pmatrix}$\n",
    "\n",
    "where \n",
    "\n",
    "- $v[0]$ represents consumption of $Na$, \n",
    "- $v[1]$ that of $HCl$, \n",
    "- $v[2]$ for $NaCl$, and \n",
    "- $v[3]$ for $H2$.\n",
    "\n",
    "This principle applied to thousands of metabolites and reactions, we get the matrix.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f7aec3d",
   "metadata": {},
   "source": [
    "### What is a GSMM\n",
    "\n",
    "Genome-scale metabolic models (GSMM or GEM for short) are formatted data files that contains all the essential information of the metabolism \n",
    "of an organism. Not only the set of reactions and metabolites, but they hold much more information such as gene associations (for turning on/off phenotypes), model compartments, flux constraints, etc. Development of comprehensive and accurate GSMMs for non-model organisms is one of the greatest bottlenecks of systems biology.\n",
    "\n",
    "Most GSMMs are published and used in the file format called SBML, which stands for **systems biology markup language**. However, the JSON file format, originally created for data transfer in web software applications, is also used for a similar purpose.\n",
    "\n",
    "### Practice\n",
    "\n",
    "The [iJO1366](http://bigg.ucsd.edu/models/iJO1366) model was developed to model the metabolism of *Escherichia coli*. In the following practice, you'll be introduced to the basics of the python library COBRApy that is used to compile these models and run very basic forms of flux-balance analysis.\n",
    "\n",
    "First start off by installing COBRApy and making necessary imports into the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "39ec458c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: cobra in c:\\users\\owner\\anaconda3\\lib\\site-packages (0.30.0)\n",
      "Requirement already satisfied: appdirs~=1.4 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (1.4.4)\n",
      "Requirement already satisfied: depinfo~=2.2 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (2.2.0)\n",
      "Requirement already satisfied: diskcache~=5.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (5.6.3)\n",
      "Requirement already satisfied: future in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (1.0.0)\n",
      "Requirement already satisfied: httpx~=0.24 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (0.28.1)\n",
      "Requirement already satisfied: numpy>=1.13 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (1.26.4)\n",
      "Requirement already satisfied: optlang~=1.8 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (1.8.3)\n",
      "Requirement already satisfied: pandas<3.0,>=1.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (2.2.3)\n",
      "Requirement already satisfied: pydantic>=1.6 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (2.11.4)\n",
      "Requirement already satisfied: python-libsbml~=5.19 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (5.21.0)\n",
      "Requirement already satisfied: rich>=8.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (14.0.0)\n",
      "Requirement already satisfied: ruamel.yaml~=0.16 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (0.18.10)\n",
      "Requirement already satisfied: swiglpk in c:\\users\\owner\\anaconda3\\lib\\site-packages (from cobra) (5.0.12)\n",
      "Requirement already satisfied: anyio in c:\\users\\owner\\anaconda3\\lib\\site-packages (from httpx~=0.24->cobra) (4.9.0)\n",
      "Requirement already satisfied: certifi in c:\\users\\owner\\anaconda3\\lib\\site-packages (from httpx~=0.24->cobra) (2025.4.26)\n",
      "Requirement already satisfied: httpcore==1.* in c:\\users\\owner\\anaconda3\\lib\\site-packages (from httpx~=0.24->cobra) (1.0.9)\n",
      "Requirement already satisfied: idna in c:\\users\\owner\\anaconda3\\lib\\site-packages (from httpx~=0.24->cobra) (3.10)\n",
      "Requirement already satisfied: h11>=0.16 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from httpcore==1.*->httpx~=0.24->cobra) (0.16.0)\n",
      "Requirement already satisfied: sympy>=1.12.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from optlang~=1.8->cobra) (1.14.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from pandas<3.0,>=1.0->cobra) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from pandas<3.0,>=1.0->cobra) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from pandas<3.0,>=1.0->cobra) (2025.2)\n",
      "Requirement already satisfied: ruamel.yaml.clib>=0.2.7 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from ruamel.yaml~=0.16->cobra) (0.2.8)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from pydantic>=1.6->cobra) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.33.2 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from pydantic>=1.6->cobra) (2.33.2)\n",
      "Requirement already satisfied: typing-extensions>=4.12.2 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from pydantic>=1.6->cobra) (4.13.2)\n",
      "Requirement already satisfied: typing-inspection>=0.4.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from pydantic>=1.6->cobra) (0.4.0)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from python-dateutil>=2.8.2->pandas<3.0,>=1.0->cobra) (1.17.0)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from rich>=8.0->cobra) (3.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from rich>=8.0->cobra) (2.19.1)\n",
      "Requirement already satisfied: mdurl~=0.1 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from markdown-it-py>=2.2.0->rich>=8.0->cobra) (0.1.2)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from sympy>=1.12.0->optlang~=1.8->cobra) (1.3.0)\n",
      "Requirement already satisfied: sniffio>=1.1 in c:\\users\\owner\\anaconda3\\lib\\site-packages (from anyio->httpx~=0.24->cobra) (1.3.1)\n"
     ]
    }
   ],
   "source": [
    "!pip3 install cobra\n",
    "\n",
    "from cobra import io, Model, Solution\n",
    "import argparse, os, sys"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bd8cb6c",
   "metadata": {},
   "source": [
    "Now, we read our model file into a `Cobra.Model` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d2b10c7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 metabolites: ['10fthf_c', '12dgr120_c', '12dgr140_c', '12dgr141_c', '12dgr160_c']\n",
      "First 5 metabolite names: ['10-Formyltetrahydrofolate', '1,2-Diacyl-sn-glycerol (didodecanoyl, n-C12:0)', '1,2-Diacyl-sn-glycerol (ditetradecanoyl, n-C14:0)', '1,2-Diacyl-sn-glycerol (ditetradec-7-enoyl, n-C14:1)', '1,2-Diacyl-sn-glycerol (dihexadecanoyl, n-C16:0)']\n",
      "First 5 reactions: ['EX_cm_e', 'EX_cmp_e', 'EX_co2_e', 'EX_cobalt2_e', 'DM_4crsol_c']\n",
      "First 5 reaction names: ['Chloramphenicol exchange', 'CMP exchange', 'CO2 exchange', 'Co2+ exchange', 'Sink needed to allow p-Cresol to leave system']\n"
     ]
    }
   ],
   "source": [
    "model: Model = io.read_sbml_model('./iJO1366.xml')\n",
    "\n",
    "five_met_ids = [met.id for met in model.metabolites[:5]]\n",
    "five_met_names = [met.name for met in model.metabolites[:5]]\n",
    "five_rxns_ids = [rxn.id for rxn in model.reactions[:5]]\n",
    "five_rxns_names = [rxn.name for rxn in model.reactions[:5]]\n",
    "print(f'First 5 metabolites: {five_met_ids}')\n",
    "print(f'First 5 metabolite names: {five_met_names}')\n",
    "print(f'First 5 reactions: {five_rxns_ids}')\n",
    "print(f'First 5 reaction names: {five_rxns_names}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b91217e3",
   "metadata": {},
   "source": [
    "The above code reads the `sbml` file into the `Model` object, and then reads out the names and ids of the first 5 reactions and metabolites in the array. To understand how data is stored in the `Model` object, we can further look into the `Metabolite` and `Reaction` classes. The output is too long, but if you want to read their documentation, you can uncomment the print statements in the below cell block."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6587f0dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra import Metabolite, Reaction\n",
    "\n",
    "# print(help(Metabolite))\n",
    "# print(help(Reaction))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c6c9e8",
   "metadata": {},
   "source": [
    "Flux-balance analysis makes a major assumption; that all metabolites are produced and consumed at the same amount. This assumption describes a steady-state (equilibrium for a set of reactions). Within this assumption, we're trying to find an arrangement of fluxes that yield the highest (maximum) flux through our target reaction (optimization objective).\n",
    "\n",
    "Let's assume we want to optimize for the **biomass** objective function. The name of this function may vary based on the GSMM and the database that was used to construct it, but the **biomass** keyword is usually in there somewhere.\n",
    "\n",
    "With the following script, we can identify the biomass reaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2ecc00e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biomass Reactions: \n",
      "- BIOMASS_Ec_iJO1366_WT_53p95M: E. coli biomass objective function (iJO1366) - WT - with 53.95 GAM estimate\n",
      "- BIOMASS_Ec_iJO1366_core_53p95M: E. coli biomass objective function (iJO1366) - core - with 53.95 GAM estimate\n"
     ]
    }
   ],
   "source": [
    "biomass_rxns = [rxn for rxn in model.reactions if 'biomass' in rxn.id.lower()]\n",
    "\n",
    "print(f'Biomass Reactions: ')\n",
    "for rxn in biomass_rxns:\n",
    "    print(f'- {rxn.id}: {rxn.name}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8fa9c7",
   "metadata": {},
   "source": [
    "Some models may have multiple biomass functions. Usually different biomass functions indicate \n",
    "\n",
    "1. the estimations via different experiments, or \n",
    "2. predictions under different medium conditions, representing different modes of the organism.\n",
    "\n",
    "As an example, in our [iGEM 2025 (Sterosaurus)](https://2025.igem.wiki/mcmasteru/) project, we used the iCre1355 model of the *C. reinhardtii* algae organism. There were 3 biomass functions representing the autotrophic, mixotrophic, and heterotrophic mode. These modes represent the growth function of the organism under different medium conditions i.e. O2 stress, CO2 stress, etc. You can find the paper for the model [here](https://onlinelibrary.wiley.com/doi/10.1111/tpj.13059).\n",
    "\n",
    "For our optimization purposes, we'll use the core biomass function `BIOMASS_Ec_iJO1366_core_53p95M`. Let's now optimize our model and get the flux results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2a3dd27d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 5 fluxes in optimal solution:\n",
      "ATPS4rpp      55.815247\n",
      "EX_h2o_e      45.619430\n",
      "CYTBO3_4pp    35.149539\n",
      "NADH16pp      31.021723\n",
      "EX_co2_e      19.675223\n",
      "Name: fluxes, dtype: float64\n",
      "\n",
      "Optimal growth rate: 0.9823718127269743 1/hr\n"
     ]
    }
   ],
   "source": [
    "biomass_obj = 'BIOMASS_Ec_iJO1366_core_53p95M'\n",
    "\n",
    "model.objective = biomass_obj\n",
    "solution: Solution = model.optimize()\n",
    "\n",
    "solution.fluxes.sort_values(ascending=False, inplace=True)\n",
    "print(f'Top 5 fluxes in optimal solution:')\n",
    "print(solution.fluxes.head(5))\n",
    "\n",
    "print(f'\\nOptimal growth rate: {solution.objective_value} 1/hr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be100268",
   "metadata": {},
   "source": [
    "In this notebook, we've explored what flux-balance analysis is, and how we can use COBRApy to load GSMMs from `sbml` files and optimize for a certain objective function.\n",
    "\n",
    "In most applications of flux-balance analysis, it is valuable to read the paper of the GSMMs model release as it contains important information relevant to our purpose.\n",
    "\n",
    "In the future\n",
    "\n",
    "### Supplementary Material\n",
    "\n",
    "You can check out last year's [GitHub repo](https://github.com/FarukEfe/McMasterU-iGEM-2025-Modelling) for flux-balance analysis. It can give you a good idea of how an fba project can be structured."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
