{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "15e37486",
   "metadata": {},
   "source": [
    "## Overview\n",
    "\n",
    "Welcome to the first notebook of the fba workshops! The learning objectives of this notebook are as follows:\n",
    "\n",
    "- Understanding flux-balance analysis\n",
    "- Understanding the use of GEMs/GSMMs in flux-balance analysis\n",
    "- Getting familiar with the COBRApy framework\n",
    "    - Importing a GSMM\n",
    "    - Finding reactions & metabolites\n",
    "    - Objective functions and running the analysis\n",
    "    - Altering reaction upper and lower bounds\n",
    "\n",
    "The take-home practice section will then try to give you intuition on:\n",
    "\n",
    "- hybrid objective functions\n",
    "- competing pathways\n",
    "- knock-outs\n",
    "\n",
    "## What is flux-balance analysis?\n",
    "\n",
    "Flux-balance analysis is a predictive modeling method that is significantly used in systems biology. It has a wide variety of applications including (but not limited to):\n",
    "\n",
    "1. Media optimization for the culture\n",
    "2. Finding knock-out targets\n",
    "3. Understanding metabolic constraints\n",
    "4. Neuroscience\n",
    "5. Biomanufacturing\n",
    "6. Cancer Research\n",
    "\n",
    "Many iGEM teams have cited [this paper](https://www.nature.com/articles/nbt.1614) to explain how flux-balance analysis works. In a nutshell, it derives from the idea that reactions and metabolites in a metabolism can be represented as a matrix by the following **example:** \n",
    "\n",
    "Let's take an arbitrary reaction:\n",
    "\n",
    "$$2Na + 2HCl \\rightarrow 2NaCl + H_2$$\n",
    "\n",
    "This reaction can be represented as a column vector:\n",
    "\n",
    "$$v = \\begin{pmatrix} -2 & -2 & 2 & 1 \\end{pmatrix}^T$$\n",
    "\n",
    "Where each item represents the coefficient consumed/produced of that metabolite in the reaction. This principle applied to thousands of metabolites and reactions, and we get a linear optimization problem formulated as $Sv = 0$ where:\n",
    "\n",
    "- $S \\in \\R^{M \\times N}$ represents the stoichiometric matrix with each column a reaction, and each row a metabolite.\n",
    "- $v \\in \\R^{N}$ represents the fluxes (rates) of reactions, usually in units $mmol/gDW/h$. \n",
    "- an assumption is made that the metabolic network is at steady-state, where each metabolite is consumed and produced at the same amount.\n",
    "\n",
    "Given a metabolite concentration $[x]$ with the corresponding coefficient $s_i$ in the $i'th$ reaction, and $t$ for time, an equivalent formulation of the problem would be $$ \\sum_{i=1}^N s_i v_i = \\frac{d[x]}{dt} = 0 $$\n",
    "\n",
    "As simple as this linear optimization looks, real use of flux-balance analysis pairs with [multiomics data](https://en.wikipedia.org/wiki/Multiomics) to carry out powerful analyses and infer more about the metabolism.\n",
    "\n",
    "### What are genome-scale metabolic models?\n",
    "\n",
    "Genome-scale metabolic models (GSMM or GEM for short) are formatted data files that contains all the essential information of the metabolism \n",
    "of an organism. Not only the set of reactions and metabolites, but they hold much more information such as gene associations (for turning on/off phenotypes), model compartments, flux constraints, etc. Development of comprehensive and accurate GSMMs for non-model organisms is one of the greatest bottlenecks of systems biology.\n",
    "\n",
    "Most GSMMs are published and used in the file format called SBML, which stands for **systems biology markup language**. However, the JSON file format, originally created for data transfer in web software applications, is also used for a similar purpose.\n",
    "\n",
    "### The COBRA Toolbox\n",
    "\n",
    "The [iJO1366](http://bigg.ucsd.edu/models/iJO1366) model was developed to mimic the metabolism of *Escherichia coli*. In the following practice, you'll be introduced to the basics of the python library COBRApy that is used to compile these models and run very basic forms of flux-balance analysis.\n",
    "\n",
    "First start off by installing COBRApy and making necessary imports into the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "39ec458c",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip3 install cobra -qqq\n",
    "\n",
    "from cobra import io, Model, Solution\n",
    "import argparse, os, sys"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bd8cb6c",
   "metadata": {},
   "source": [
    "Now, we read our model file into a `Cobra.Model` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "d2b10c7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 metabolites: \n",
      "(1) ID: 10fthf_c - Name: 10-Formyltetrahydrofolate\n",
      "(2) ID: 12dgr120_c - Name: 1,2-Diacyl-sn-glycerol (didodecanoyl, n-C12:0)\n",
      "(3) ID: 12dgr140_c - Name: 1,2-Diacyl-sn-glycerol (ditetradecanoyl, n-C14:0)\n",
      "(4) ID: 12dgr141_c - Name: 1,2-Diacyl-sn-glycerol (ditetradec-7-enoyl, n-C14:1)\n",
      "(5) ID: 12dgr160_c - Name: 1,2-Diacyl-sn-glycerol (dihexadecanoyl, n-C16:0)\n",
      "\n",
      "First 5 reactions: \n",
      "(1) ID: EX_cm_e - Name: Chloramphenicol exchange\n",
      "(2) ID: EX_cmp_e - Name: CMP exchange\n",
      "(3) ID: EX_co2_e - Name: CO2 exchange\n",
      "(4) ID: EX_cobalt2_e - Name: Co2+ exchange\n",
      "(5) ID: DM_4crsol_c - Name: Sink needed to allow p-Cresol to leave system\n"
     ]
    }
   ],
   "source": [
    "model: Model = io.read_sbml_model('./iJO1366.xml')\n",
    "\n",
    "five_mets = [f'\\n({i+1}) ID: {met.id} - Name: {met.name}' for i, met in enumerate(model.metabolites[:5])]\n",
    "five_rxns = [f'\\n({i+1}) ID: {rxn.id} - Name: {rxn.name}' for i, rxn in enumerate(model.reactions[:5])]\n",
    "print(f'First 5 metabolites: {''.join(five_mets)}\\n')\n",
    "print(f'First 5 reactions: {''.join(five_rxns)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b91217e3",
   "metadata": {},
   "source": [
    "The above code reads the `sbml` file into the `Model` object, and then reads out the names and ids of the first 5 reactions and metabolites in the array. To understand how data is stored in the `Model` object, we can further look into the `Metabolite` and `Reaction` classes.\n",
    "\n",
    "Here's a neater way to display the model information on jupyter notebook, which neatly renders as a table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "c1b14a3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iJO1366</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>1fde355b860</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1805</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2583</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1367</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_Ec_iJO1366_core_53p95M - 1.0*BIOMASS_Ec_iJO1366_core_53p95M_reverse_5c8b1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iJO1366 at 0x1fde355b860>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5db344b4",
   "metadata": {},
   "source": [
    "The output is too long, but if you want to read the documentation of the `Metabolite` and `Reaction` objects, feel free to uncomment the below lines. Alternatively, you can visit the [COBRApy documentation page](https://cobrapy.readthedocs.io/en/latest/) for more information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "6587f0dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra import Metabolite, Reaction\n",
    "\n",
    "# print(help(Metabolite))\n",
    "# print(help(Reaction))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c6c9e8",
   "metadata": {},
   "source": [
    "As previously discussed, flux-balance analysis makes a major assumption; that all metabolites are produced and consumed at the same amount. This assumption describes a steady-state (equilibrium for a set of reactions). Within this assumption, we're trying to find an arrangement of fluxes that yield the highest (maximum) flux through our target reaction (objective).\n",
    "\n",
    "Let's assume we want to optimize for the **biomass** objective function. The name of this function may vary based on the GSMM and the database that was used to construct it, but the **biomass** keyword is usually in there somewhere.\n",
    "\n",
    "With the following script, we can identify the biomass reaction by looking for the **biomass** keyword in all reaction IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "2ecc00e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biomass Reactions: \n",
      "- BIOMASS_Ec_iJO1366_WT_53p95M: E. coli biomass objective function (iJO1366) - WT - with 53.95 GAM estimate\n",
      "- BIOMASS_Ec_iJO1366_core_53p95M: E. coli biomass objective function (iJO1366) - core - with 53.95 GAM estimate\n"
     ]
    }
   ],
   "source": [
    "biomass_rxns = [rxn for rxn in model.reactions if 'biomass' in rxn.id.lower()]\n",
    "\n",
    "print(f'Biomass Reactions: ')\n",
    "for rxn in biomass_rxns:\n",
    "    print(f'- {rxn.id}: {rxn.name}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8fa9c7",
   "metadata": {},
   "source": [
    "Some models may have multiple biomass functions. Usually different biomass functions indicate \n",
    "\n",
    "1. the estimations via different experiments, or \n",
    "2. predictions under different medium conditions, representing different modes of the organism.\n",
    "\n",
    "As an example, in our [iGEM 2025 (Sterosaurus)](https://2025.igem.wiki/mcmasteru/) project, we used the iCre1355 model of the *C. reinhardtii* algae organism. There were 3 biomass functions representing the autotrophic, mixotrophic, and heterotrophic mode. These modes represent the growth function of the organism under different medium conditions i.e. O2 stress, CO2 stress, etc. You can find the paper for the model [here](https://onlinelibrary.wiley.com/doi/10.1111/tpj.13059).\n",
    "\n",
    "For our optimization purposes, we'll use the core biomass function `BIOMASS_Ec_iJO1366_core_53p95M`. Let's now optimize our model and get the flux results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "2a3dd27d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 5 fluxes in optimal solution:\n",
      "ATPS4rpp      55.815247\n",
      "EX_h2o_e      45.619430\n",
      "CYTBO3_4pp    35.149539\n",
      "NADH16pp      31.021723\n",
      "EX_co2_e      19.675223\n",
      "Name: fluxes, dtype: float64\n",
      "\n",
      "Optimal growth rate: 0.9823718127269743 1/hr\n"
     ]
    }
   ],
   "source": [
    "biomass_obj = 'BIOMASS_Ec_iJO1366_core_53p95M'\n",
    "\n",
    "model.objective = biomass_obj\n",
    "solution: Solution = model.optimize()\n",
    "\n",
    "solution.fluxes.sort_values(ascending=False, inplace=True)\n",
    "print(f'Top 5 fluxes in optimal solution:')\n",
    "print(solution.fluxes.head(5))\n",
    "\n",
    "print(f'\\nOptimal growth rate: {solution.objective_value} 1/hr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf0848e3",
   "metadata": {},
   "source": [
    "Additionally, reactions can be upper and lower bounded within the model. We can do this by reaching a reaction of our choice within the model and change its `upper_bound` and `lower_bound` parameters. Let's try to do this on the medium uptake of our model.\n",
    "\n",
    "To find a reaction of interest, we may think we need to iterate through every reaction in the model. However, there's simple ways in which we can reduce the workload. For instance, the cobra `Model` object is known to keep record of the compartments of each metabolite. We can first write a script to print out the compartments available in the model, and then filter only for the exchange reactions that have extracellular metabolites as reactants.\n",
    "\n",
    "We start by listing out the compartments of our model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "379e387f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'c': 'cytosol', 'e': 'extracellular space', 'p': 'periplasm'}"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.compartments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bfd66a0",
   "metadata": {},
   "source": [
    "Notice that the `'e': 'extracellular space'` compartment fits our use case. Now, let's print out each metabolite found in this compartment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "5c1c33d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('co2_e', 'CO2 CO2'),\n",
       " ('h2o2_e', 'Hydrogen peroxide'),\n",
       " ('kdo2lipid4_e', 'KDO(2)-lipid IV(A)'),\n",
       " ('so2_e', 'Sulfur dioxide'),\n",
       " ('no2_e', 'Nitrite'),\n",
       " ('o2_e', 'O2 O2'),\n",
       " ('o2s_e', 'Superoxide anion')]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Query for all metabolites with 'o2' in their ID or name\n",
    "mets = model.metabolites.query('.*o2.*')\n",
    "# Print only those in the extracellular compartment\n",
    "output = []\n",
    "for met in mets:\n",
    "    if met.compartment != 'e':\n",
    "        continue\n",
    "    output.append((met.id, met.name))\n",
    "\n",
    "output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e900728",
   "metadata": {},
   "source": [
    "Now, let's try to constrain the model based on its oxygen uptake. We now find the oxygen uptake reaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "ee506ae4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- EX_o2_e: O2 exchange: -1000.0 to 1000.0 mmol/gDW/hr\n",
      "- O2tex: Oxygen transport via diffusion (extracellular to periplasm): -1000.0 to 1000.0 mmol/gDW/hr\n"
     ]
    }
   ],
   "source": [
    "o2e_id, o2e_name = output[5]\n",
    "\n",
    "rxns = []\n",
    "for rxn in model.reactions:\n",
    "    if o2e_id in [met.id for met, coef in rxn.metabolites.items() if coef < 0]:\n",
    "        rxns.append(rxn)\n",
    "\n",
    "for r in rxns:\n",
    "    print(f'- {r.id}: {r.name}: {r.lower_bound} to {r.upper_bound} mmol/gDW/hr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d31549b",
   "metadata": {},
   "source": [
    "In the context of a metabolic model, a range of -1000 to 1000 is practically infinite. Let's set this much lower. The reason why we take the copy of the original model in the below code is because `Model` and `Reaction` behave as reference types, and the variable `oxygen_uptake_rxn` can mutate the original object in the model. Making a copy prevents us from overwriting the original data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "9f52c917",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Optimal growth rate: 0.9823718127269705 1/hr\n",
      "\n",
      "Optimal growth rate: 0.49154514798795895 1/hr\n"
     ]
    }
   ],
   "source": [
    "# Original optimization\n",
    "solution = model.optimize()\n",
    "\n",
    "print(f'\\nOptimal growth rate: {solution.objective_value} 1/hr')\n",
    "\n",
    "# With new boundaries\n",
    "copy_model = model.copy()\n",
    "oxygen_uptake_rxn = copy_model.reactions.get_by_id('EX_o2_e')\n",
    "oxygen_uptake_rxn.lower_bound = -5.0\n",
    "oxygen_uptake_rxn.upper_bound = 5.0 \n",
    "\n",
    "solution = copy_model.optimize()\n",
    "\n",
    "print(f'\\nOptimal growth rate: {solution.objective_value} 1/hr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e368a949",
   "metadata": {},
   "source": [
    "Now that we've covered some of the basics of COBRApy, we can get started with some take-home practice questions.\n",
    "\n",
    "## Take-Home Practice\n",
    "\n",
    "#### Part 1: Hybrid Objective for Optimization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be100268",
   "metadata": {},
   "source": [
    "In this notebook, we've explored what flux-balance analysis is, and how we can use COBRApy to load GSMMs from `sbml` files and optimize for a certain objective function.\n",
    "\n",
    "In most applications of flux-balance analysis, it is valuable to read the paper of the GSMMs model release as it contains important information relevant to our purpose.\n",
    "\n",
    "In the future\n",
    "\n",
    "### Supplementary Material\n",
    "\n",
    "You can check out last year's [GitHub repo](https://github.com/FarukEfe/McMasterU-iGEM-2025-Modelling) for flux-balance analysis. It can give you a good idea of how an fba project can be structured.\n",
    "\n",
    "As previously mentioned, feel free to check out the [COBRApy documentation](https://cobrapy.readthedocs.io/en/latest/) on the use of the library.\n",
    "\n",
    "If you're curious, you can also check out [our software repository](https://gitlab.igem.org/2025/software-tools/mcmasteru) from iGEM 2025, where the flux-balance analysis code can be found under the `fba-codebase` folder."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa5d9c49",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
